docs/spec/fastsync/fastsync.md

```diff
@@ -0,0 +1,1164 @@
+nodes called *peers* that it knows of.  The full node uses the peers
+to read blocks of the Tendermint blockchain (in a safe way, that is,
+it checks the soundness conditions), until it has read the most recent
+block and then terminates.
+
+
+## Sequential Problem statement
+
+
+
+*Fastsync* gets as input a block of height *h* and the corresponding
+application state *s* that corresponds to the block and state of that
+height of the blockchain, and produces
+as output (i) a list *L* of blocks starting at height *h* to some height
+*terminationHeight*, and (ii) the application state when applying the
+transactions of the list *L* to *s*. Fastsync has to satisfy the following
+properties [FS-SEQ-?]:
+
+
+#### **[FS-SEQ-LIVE]**: 
+*Fastsync* eventually terminates.
+
+ 
+#### **[FS-SEQ-TERM]**:
+Let *bh* be the height of the blockchain at the time *Fastsync*
+starts. By assumption we have *bh >= h*.
+When *Fastsync* terminates, it outputs a list of all blocks from
+height *h* to some height *terminationHeight >= bh - 1*.
+
+> The above property is independent of how many blocks are added to the
+> blockchain while Fastsync is running. It links the target height to the
+> initial state. If Fastsync has to catch-up many blocks, it would be
+> better to link the target height to a time close to the
+> termination. This is capture by the following specification:
+
+#### **[FS-SEQ-TERM-SYNC]**:
+Let *eh* be the height of the blockchain at the time *Fastsync*
+terminates.  There is a constant *D >= 1* such that when *Fastsync*
+terminates, it outputs a list of all blocks from height *h* to some
+height *terminationHeight >= eh - D*.
+
+
+#### **[FS-SEQ-INV]**:
+Upon termination, the application state is the one that corresponds to
+the blockchain at height *terminationHeight*.
+
+> As the block of height *h + 1* is needed to verify the block of
+> height *h* we highlight th following clarification of the
+> termination height.
+
+
+#### **[FS-SEQ-HEIGHT]**: 
+The returned value *terminationHeight* is the height of the block with the largest
+height that could be verified. In order to do so, *Fastsync* needs the
+Commit of the block at height  *terminationHeight + 1* in the blockchain.
+
+
+# Part II - Protocol view
+
+## Environment/Assumptions/Incentives
+
+
+#### **[FS-A-NODE]**:
+We consider a node *FS* that performs *Fastsync*.
+
+#### **[FS-A-PEER-IDS]**:
+*FS* has access to a set *peerIDs* of IDs (public keys) of peers (full
+     nodes). During the execution of *Fastsync*, another protocol (outside
+     of this specification) may add new IDs to *peerIDs*.
+
+
+
+#### **[FS-A-PEER]**:
+Peers can be faulty, and we do not make any assumptions about the number or
+ratio of correct/faulty nodes. Faulty processes may be Byzantine
+according to [**[TMBC-AUTH-BYZ]**][TMBC-Auth-Byz-link].
+
+#### **[FS-A-VAL]**:
+The system satisfies [**[TMBC-AUTH-BYZ]**][TMBC-Auth-Byz-link] and [**[TMBC-FM-2THIRDS]**][TMBC-FM-2THIRDS-link]. Thus, there is a
```

Comment:
This is not really clear in my view. We have a node and peers, and I guess that with this property we want to make a link to validators and Tendermint blockchain. I am not sure what is the easiest way to introduce this without going into details of explaining validators and difference between validators and full nodes. From the fast sync view point the whole blockchain business can be seen as application specific stuff, and it is only relevant as part of block validation in the proposed solution.

Reply:
If we want to say what blocks to accept and execute, we have to state that there is a reference block chain that has these blocks. It is not clear how to specify reasonable invariants without having that reference. It is the validity property of Fastsync.   
Later, in our suggestions to use the lightclient-like verification (at the end of the document) we need this as it is based on the tendermint failure model.